cEP-0022.md

```diff
@@ -0,0 +1,497 @@
+- *"Green"* values for settings are added to `.coafile`
+
+# Acting upon the newly available metadata
+
+The metadata has been generated in the following files:
+- [`.quickfile`](#quickfile)
+- [`QUICKDATA.yaml`](#quickdatayaml)
+- [`QUICKINFO.yaml`](#quickinfoyaml)
+
+and the following classes:
+- [`Quickclass`](#quickclass)
+
+## `.quickfile`
+
+Will have 3 sections as created in `.coafile`
+- **[severe]**<a name="severe"></a>
+  This will contain the **Type 1** and **Type 3** dropped settings along with their
+  values and weights for each value.
+- **[amenable]**<a name="amenable"></a>
+  This will contain the **Type 2** dropped settings along with their weights and
+  values.
+- **[permanent]**<a name="permanent"></a>
+  This will contain values to settings that are obtained when the user answers
+  the questions prepared by `coala-quickstart`. Special string may be used
+  to represent confirmed settings that need to be dropped.
+
+`coala-quickstart` will always ask at the end of the run
+whether the user is interested in answering
+some questions which will lead to more secure creation of config files. If the
+user choses to answer them, `coala-quickstart` will provide the user with a
+bunch of questions from the sections **[severe]** an **[amenable]** asking
+them whether an inconsistency detected is a mistake in the code-base or there
+are no style rules followed by the project regarding the particular setting.
+The answered bear-settings
+will be moved to the **[permanent]** section.
+
+These set of questions can be invoked directly without a `--green-mode` run
+or not provided at all by providing special tags.
+[*(Check out the tags section for details)*](#list-of-tags-introduced)
+
+The user may also be asked to provide a `SEVERITY` value which will be mapped
+to the weights of values store in the `.quickfile`. This mapping will be
+decided after testing the `--green-mode` on some projects. All inconsistencies
+with lesser `SEVERITY` value than the one entered will be displayed.
+
+In case the user wants to check exactly what lines produce the inconsistencies,
+coala may be ran again with just that bear and setting value to show the
+user the erroneous lines in the code-base.
+
+All the necessary information regarding the `SEVERITY` value will be provided by
+the initial question.
+
+## `QUICKDATA.yaml`
+
+Will contain the data generated by QuickstartBear as described earlier used to
+guess
+**Type 2** settings
+
+## `QUICKINFO.yaml`
+
+Will contain the accepted combinations of values of bear settings along with the
+project url. These files will be uploaded as Pull Requests to a repository
+created specifically to store this data.
+
+Encryption will be applied so that any user cannot mess with the data generated
+by `coala-quickstart`, thus making Pull Request for junk data
+and will be available to view only when it has been
+merged to the repository.
+
+Tools like gitmate-plugins to accept valid Pull Requests automatically and
```

Comment:
Please check with gitmate team and get them to approve your cEP if you refer to them.

They already worked on a tool like this during gci '17

Reply:
>I dont see enough emphasis on each file having different settings, or specific files being detected as completely broken and ignored from various setting sections.

Like I have mentioned in the cEP, whenever we assume a setting and see incompliant code with that setting, we always have 2 options:
* either to assume it as an inconsistency and add it to `.quickfile`
* or two assume it a code style and add an ignore, make a different section for that file and all that stuff

but how do we decide which one of the two it is? maybe we decide a threshold of inconsistencies below which they are considered as mistakes and after which we assume them as code styles . But to calculate these thresholds we again don't have an efficient way to do it (you might remember me mentioning asking things like your competency level in the proposal for this). This is already pointed out by Satwik (may be done later but for now we have dropped this calculation of threshold).

I have made some assumptions in my proposal that it is highly unlikely for an org to follow different kinds of code styles for different file types

>+There can be endless number of possibilities and assumptions:
 +Files having similar kind of names can be having different codestyles.
 +Files in a given directory can be following different codestyles.
 +Different parts of files maybe even following a different codestyle depending
 +on function name which can also be handled by placing `ignore`s at
 +appropriate places in the file dict.

We don't know for now what we should be doing so instead I am storing all the meta data I can, and on the basis of that assumption creating green config files. If we get different trends in the orgs we choose or even later on, we adapt, till then there is the point about the annoying question asking to answer the user some questions about the project to settle these ambiguities.

----
----

>Many bools are tertinery, includes None.
Are they excluded from type 1?

let me give you an example:
take for eg. the setting: use_spaces: bool from SpaceConsistencyBear
so the project can either use spaces and get a green setting, or use tabs and again get a different setting, but if its a mixture, the setting has to be dropped (more on this as a reply on your next comment) i.e. takes the value None

----
----

>These pycodestyle settings are a great example of a problem, and this approach isnt the ideal solution for them, as it relies on existing config.
These settings are effectively a set of bools.
See discussion in coala/ast about these settings.
I believe the info parsing component of quickstart is a green mode helper, but it shouldnt be the only way a setting can become green.

couldn't get much from coala/ast about these, and I couldn't get why you thinks this is not an ideal solution because fo dependency on existing config files. If the user hasn't specified these settings in the existing config files it means they are not meant to be used, it may happen that the user may be running this on a local project and the config file (global) ones may be taken up by quickstart but that is more of the user's problem, we can give out additional output by quickstart about all the config files we are using to determine these settings. if we don't find mention of these settings in the config files, we drop them (Assume to be false), but if we find the mention of these settings int he config file, we parse the file and ignore and select the error messages/warnings provided by these settings accordingly.

----
----

>Please check with gitmate team and get them to approve your cEP if you refer to them.
They already worked on a tool like this during gci '17

Initial pull requests will be merged manually, this is more of enhancing/automating the metadata 
collection part, which will be worked upon towards the end of the project. Will get in touch with the gitmate team ASAP